pkg/solver/controller.go

```diff
@@ -403,6 +429,62 @@ func (controller *SolverController) cancelExpiredJobs(ctx context.Context) error
 	return nil
 }
 
+func (controller *SolverController) cancelExpiredJob(ctx context.Context, jobOffer data.JobOfferContainer, expirationState string) (jobOfferID string, dealID string) {
+	ctx, span := controller.tracer.Start(ctx, "cancel_expired_job",
+		trace.WithAttributes(attribute.String("job_offer.id", jobOffer.ID)),
+		trace.WithAttributes(attribute.String("job_offer.deal_id", jobOffer.DealID)),
+		trace.WithAttributes(attribute.String("job_offer.job_creator", jobOffer.JobCreator)),
+		trace.WithAttributes(attribute.String("job_offer.state", data.GetAgreementStateString(jobOffer.State))),
+		trace.WithAttributes(attribute.String("expiration_state", expirationState)),
+	)
+	defer span.End()
+
+	// Cancel pre-match expired job offer
+	if jobOffer.DealID == "" {
+		_, err := controller.updateJobOfferState(jobOffer.ID, jobOffer.DealID, data.GetAgreementStateIndex(expirationState))
+		if err != nil {
+			controller.log.Error().Func(system.AddTraceContext(span)).Err(err).Msg("update expired job offer state failed")
+			span.SetStatus(codes.Error, "update expired job offer state failed")
+			span.RecordError(err)
+		}
+
+		return jobOffer.ID, ""
+	}
+
+	// Cancel expired job offer, resource offer, and deal
+	deal, err := controller.updateDealState(jobOffer.DealID, data.GetAgreementStateIndex(expirationState))
+	if err != nil {
+		controller.log.Error().Func(system.AddTraceContext(span)).Err(err).Msg("update expired deal state failed")
```

Comment:
Let's add some contextual information from the JobOffer in here to help with debugging